File: source/discovery.ts
/**
 * @copyright Copyright (c) 2026 GUIHO Technologies as represented by Cristóvão GUIHO. All Rights Reserved.
 */

import { readDirectory as readdir, statPath as stat } from './runtime/fs.js'
import { basename, dirname, joinPath as join, relativePath as relative, resolvePath as resolve } from './runtime/path.js'
import type { XDocsConfig, XDocsFile, XDocsMarkdownDocument, XDocsScanResult } from './types.js'
import { parseXDocsFile } from './metadata.js'

const XDOCS_DESCRIPTOR_EXTENSION = '.xdocs.md'
const ROOT_XDOCS_FILENAME = 'XDOCS.md'

/** Scan the project for xdocs descriptor files and sibling Markdown documents. */
export const scanProject = async (config: XDocsConfig): Promise<XDocsScanResult> => {
  const xdocsFiles: XDocsFile[] = []
  const markdownDocuments: XDocsMarkdownDocument[] = []
  const markdownDocumentsByDirectory = new Map<string, XDocsMarkdownDocument[]>()
  const allDirectories: string[] = []
  let totalFiles = 0

  await walkDirectory(config.cwd, config, async (filePath, _dir) => {
    totalFiles += 1

    if (isXDocsDescriptorFile(filePath)) {
      const file = await parseXDocsFile(filePath, config.cwd)
      xdocsFiles.push(file)
      return
    }

    if (isPlainMarkdownDocument(filePath)) {
      const document = createMarkdownDocument(filePath, config.cwd)
      markdownDocuments.push(document)
      addMarkdownDocument(markdownDocumentsByDirectory, document.directory, document)
    }
  }, allDirectories)

  // The root XDOCS.md is always recognized
  const rootXDocs = join(config.cwd, ROOT_XDOCS_FILENAME)
  const rootAlreadyFound = xdocsFiles.some((f) => f.path === rootXDocs)

  if (!rootAlreadyFound) {
    try {
      const rootStat = await stat(rootXDocs)
      if (rootStat.isFile()) {
        const file = await parseXDocsFile(rootXDocs, config.cwd)
        xdocsFiles.push(file)
      }
    } catch {
      // Root XDOCS.md does not exist, that's fine
    }
  }

  enrichDescriptorFiles(xdocsFiles, markdownDocumentsByDirectory)

  const coveredDirectories = new Set<string>()
  if (xdocsFiles.some((file) => file.path === rootXDocs)) coveredDirectories.add(config.cwd)
  for (const file of xdocsFiles) {
    if (file.path === rootXDocs) continue
    if (file.valid) coveredDirectories.add(file.directory)
  }

  const uncoveredPaths = allDirectories.filter((dir) => !coveredDirectories.has(dir))

  return {
    totalFiles,
    totalDirectories: allDirectories.length,
    totalMarkdownDocuments: markdownDocuments.length,
    coveredDirectories: coveredDirectories.size,
    uncoveredDirectories: uncoveredPaths.length,
    xdocsFiles,
    markdownDocuments,
    uncoveredPaths,
  }
}

/** Check if a file path is a root index or xdocs descriptor file. */
export const isXDocsFile = (filePath: string, _extensions: string[] = [XDOCS_DESCRIPTOR_EXTENSION]): boolean =>
  basename(filePath) === ROOT_XDOCS_FILENAME || isXDocsDescriptorFile(filePath)

/** Check if a file path is an xdocs module descriptor. */
export const isXDocsDescriptorFile = (filePath: string): boolean =>
  basename(filePath).toLowerCase().endsWith(XDOCS_DESCRIPTOR_EXTENSION)

/** Check if a file path is a companion Markdown document, not an xdocs descriptor. */
export const isPlainMarkdownDocument = (filePath: string): boolean => {
  const name = basename(filePath)
  const lower = name.toLowerCase()
  return lower.endsWith('.md') && lower !== 'xdocs.md' && !isXDocsDescriptorFile(filePath)
}

/** Recursively walk a directory, skipping excluded directories. */
const walkDirectory = async (
  dir: string,
  config: XDocsConfig,
  onFile: (filePath: string, directory: string) => Promise<void>,
  allDirectories: string[],
): Promise<void> => {
  allDirectories.push(dir)

  let entries

  try {
    entries = await readdir(dir, { withFileTypes: true })
  } catch {
    return
  }

  for (const entry of entries) {
    const fullPath = join(dir, entry.name)

    if (entry.isDirectory()) {
      if (isExcluded(entry.name, config.scan.exclude)) continue
      await walkDirectory(fullPath, config, onFile, allDirectories)
    } else if (entry.isFile()) {
      await onFile(fullPath, dir)
    }
  }
}

/** Check if a directory name is in the exclude list. */
const isExcluded = (name: string, exclude: string[]): boolean =>
  exclude.includes(name) || name.startsWith('.')

/** Scan a specific directory (not recursive) for xdocs descriptors. */
export const scanDirectory = async (dirPath: string, config: XDocsConfig): Promise<XDocsFile[]> => {
  const files: XDocsFile[] = []
  const markdownDocumentsByDirectory = new Map<string, XDocsMarkdownDocument[]>()
  let entries

  try {
    entries = await readdir(dirPath, { withFileTypes: true })
  } catch {
    return files
  }

  for (const entry of entries) {
    if (!entry.isFile()) continue
    const fullPath = join(dirPath, entry.name)

    if (isXDocsDescriptorFile(fullPath) || isRootXDocsFile(fullPath, config.cwd)) {
      const file = await parseXDocsFile(fullPath, config.cwd)
      files.push(file)
      continue
    }

    if (isPlainMarkdownDocument(fullPath)) {
      const document = createMarkdownDocument(fullPath, config.cwd)
      addMarkdownDocument(markdownDocumentsByDirectory, document.directory, document)
    }
  }

  enrichDescriptorFiles(files, markdownDocumentsByDirectory)
  return files
}

/** List all files in a directory (non-recursive). */
export const listDirectoryFiles = async (dirPath: string, config: XDocsConfig): Promise<string[]> => {
  const result: string[] = []
  let entries

  try {
    entries = await readdir(dirPath, { withFileTypes: true })
  } catch {
    return result
  }

  for (const entry of entries) {
    if (!entry.isFile()) continue
    if (isExcluded(entry.name, config.scan.exclude)) continue
    result.push(relative(config.cwd, join(dirPath, entry.name)))
  }

  return result
}

const isRootXDocsFile = (filePath: string, cwd: string): boolean =>
  resolve(filePath) === resolve(cwd, ROOT_XDOCS_FILENAME)

const createMarkdownDocument = (filePath: string, cwd: string): XDocsMarkdownDocument => ({
  path: filePath,
  relativePath: relative(cwd, filePath),
  directory: dirname(filePath),
  name: basename(filePath),
})

const addMarkdownDocument = (
  documentsByDirectory: Map<string, XDocsMarkdownDocument[]>,
  directory: string,
  document: XDocsMarkdownDocument,
) => {
  const documents = documentsByDirectory.get(directory) ?? []
  documents.push(document)
  documentsByDirectory.set(directory, documents)
}

const enrichDescriptorFiles = (
  files: XDocsFile[],
  documentsByDirectory: Map<string, XDocsMarkdownDocument[]>,
) => {
  const descriptorsByDirectory = new Map<string, XDocsFile[]>()

  for (const file of files) {
    if (!isXDocsDescriptorFile(file.path)) continue
    const descriptors = descriptorsByDirectory.get(file.directory) ?? []
    descriptors.push(file)
    descriptorsByDirectory.set(file.directory, descriptors)
  }

  for (const descriptors of descriptorsByDirectory.values()) {
    if (descriptors.length <= 1) continue

    for (const file of descriptors) {
      file.errors.push('Multiple xdocs descriptors found in this directory. Keep exactly one named "*.xdocs.md" file per directory.')
      file.valid = false
    }
  }

  for (const file of files) {
    if (!isXDocsDescriptorFile(file.path)) continue

    file.documents = documentsByDirectory.get(file.directory) ?? []
    validateDocumentReferences(file)
    if (file.errors.length > 0) file.valid = false
  }
}

const validateDocumentReferences = (file: XDocsFile) => {
  if (!file.metadata) return

  const actualDocuments = new Set(file.documents.map((document) => document.name))
  const declaredDocuments = file.metadata.documents

  for (const document of file.documents) {
    if (Object.hasOwn(declaredDocuments, document.name)) continue
    file.errors.push(`Undocumented Markdown document: "${document.name}" must be listed in the "documents" metadata map.`)
  }

  for (const name of Object.keys(declaredDocuments)) {
    if (!isPlainMarkdownDocumentName(name)) {
      file.errors.push(`Invalid document entry: "${name}" must be a sibling plain "*.md" filename, not an xdocs descriptor or path.`)
      continue
    }

    if (!actualDocuments.has(name)) {
      file.errors.push(`Missing Markdown document: "${name}" is listed in metadata but does not exist beside the descriptor.`)
    }
  }
}

const isPlainMarkdownDocumentName = (name: string): boolean => {
  const lower = name.toLowerCase()
  return !name.includes('/') && !name.includes('\\') && lower.endsWith('.md') && lower !== 'xdocs.md' && !lower.endsWith(XDOCS_DESCRIPTOR_EXTENSION)
}

File: internal/xdocs/discovery.go
package xdocs

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/CGuiho/xdocs/internal/config"
)

const (
	descriptorSuffix = ".xdocs.md"
	rootFilename     = "XDOCS.md"
)

func ScanProject(cfg config.Config) (ScanResult, error) {
	result := ScanResult{
		XDocsFiles:        []File{},
		MarkdownDocuments: []MarkdownDocument{},
		UncoveredPaths:    []string{},
	}
	documentsByDirectory := map[string][]MarkdownDocument{}
	var directories []string

	err := filepath.WalkDir(cfg.CWD, func(path string, entry os.DirEntry, walkErr error) error {
		if walkErr != nil {
			if entry != nil && entry.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}
		if entry.IsDir() {
			if path != cfg.CWD && excluded(entry.Name(), cfg.Exclude) {
				return filepath.SkipDir
			}
			directories = append(directories, path)
			return nil
		}
		result.TotalFiles++
		switch {
		case IsDescriptor(path):
			result.XDocsFiles = append(result.XDocsFiles, ParseFile(path, cfg.CWD))
		case IsPlainMarkdown(path):
			document := MarkdownDocument{
				Path:         path,
				RelativePath: slashRelative(cfg.CWD, path),
				Directory:    filepath.Dir(path),
				Name:         filepath.Base(path),
			}
			result.MarkdownDocuments = append(result.MarkdownDocuments, document)
			documentsByDirectory[document.Directory] = append(documentsByDirectory[document.Directory], document)
		}
		return nil
	})
	if err != nil {
		return ScanResult{}, err
	}

	rootPath := filepath.Join(cfg.CWD, rootFilename)
	if content, readErr := os.ReadFile(rootPath); readErr == nil {
		result.XDocsFiles = append(result.XDocsFiles, File{
			Path: rootPath, RelativePath: rootFilename, Directory: cfg.CWD,
			Documents: []MarkdownDocument{}, Body: string(content), Valid: false, Errors: []string{},
		})
	}
	sort.Slice(result.XDocsFiles, func(i, j int) bool { return result.XDocsFiles[i].RelativePath < result.XDocsFiles[j].RelativePath })
	sort.Slice(result.MarkdownDocuments, func(i, j int) bool {
		return result.MarkdownDocuments[i].RelativePath < result.MarkdownDocuments[j].RelativePath
	})
	enrichFiles(result.XDocsFiles, documentsByDirectory)

	covered := map[string]bool{}
	if info, err := os.Stat(rootPath); err == nil && info.Mode().IsRegular() {
		covered[cfg.CWD] = true
	}
	for _, file := range result.XDocsFiles {
		if file.Valid {
			covered[file.Directory] = true
		}
	}
	for _, directory := range directories {
		if !covered[directory] {
			result.UncoveredPaths = append(result.UncoveredPaths, slashRelative(cfg.CWD, directory))
		}
	}
	result.TotalDirectories = len(directories)
	result.TotalMarkdownDocuments = len(result.MarkdownDocuments)
	result.CoveredDirectories = len(covered)
	result.UncoveredDirectories = len(result.UncoveredPaths)
	return result, nil
}

func IsDescriptor(path string) bool {
	return strings.HasSuffix(strings.ToLower(filepath.Base(path)), descriptorSuffix)
}

func IsPlainMarkdown(path string) bool {
	name := filepath.Base(path)
	lower := strings.ToLower(name)
	return strings.HasSuffix(lower, ".md") && !strings.EqualFold(name, rootFilename) && !IsDescriptor(path)
}

func excluded(name string, values []string) bool {
	if strings.HasPrefix(name, ".") {
		return true
	}
	for _, value := range values {
		if name == value {
			return true
		}
	}
	return false
}

func enrichFiles(files []File, documents map[string][]MarkdownDocument) {
	byDirectory := map[string][]int{}
	for index := range files {
		if !IsDescriptor(files[index].Path) {
			continue
		}
		byDirectory[files[index].Directory] = append(byDirectory[files[index].Directory], index)
	}
	for _, indexes := range byDirectory {
		if len(indexes) <= 1 {
			continue
		}
		for _, index := range indexes {
			files[index].Errors = append(files[index].Errors, `Multiple xdocs descriptors found in this directory. Keep exactly one named "*.xdocs.md" file per directory.`)
			files[index].Valid = false
		}
	}
	for index := range files {
		file := &files[index]
		if !IsDescriptor(file.Path) {
			continue
		}
		file.Documents = append([]MarkdownDocument(nil), documents[file.Directory]...)
		sort.Slice(file.Documents, func(i, j int) bool { return file.Documents[i].Name < file.Documents[j].Name })
		validateDocumentReferences(file)
		file.Valid = file.Metadata != nil && len(file.Errors) == 0
	}
}

func validateDocumentReferences(file *File) {
	if file.Metadata == nil {
		return
	}
	actual := map[string]bool{}
	for _, document := range file.Documents {
		actual[document.Name] = true
		if _, ok := file.Metadata.Documents[document.Name]; !ok {
			file.Errors = append(file.Errors, fmt.Sprintf(`Undocumented Markdown document: "%s" must be listed in the "documents" metadata map.`, document.Name))
		}
	}
	for name := range file.Metadata.Documents {
		if !validPlainMarkdownName(name) {
			file.Errors = append(file.Errors, fmt.Sprintf(`Invalid document entry: "%s" must be a sibling plain "*.md" filename, not an xdocs descriptor or path.`, name))
		} else if !actual[name] {
			file.Errors = append(file.Errors, fmt.Sprintf(`Missing Markdown document: "%s" is listed in metadata but does not exist beside the descriptor.`, name))
		}
	}
}

func validPlainMarkdownName(name string) bool {
	lower := strings.ToLower(name)
	return filepath.Base(name) == name && !strings.ContainsAny(name, `/\`) &&
		strings.HasSuffix(lower, ".md") && !strings.EqualFold(name, rootFilename) &&
		!strings.HasSuffix(lower, descriptorSuffix)
}

func inScope(path, target string) bool {
	relative, err := filepath.Rel(target, path)
	return err == nil && relative != ".." && !strings.HasPrefix(relative, ".."+string(filepath.Separator))
}
